src/supervised/margin.jl

```diff
@@ -379,3 +379,121 @@ isstrictlyconvex(::L2MarginLoss) = true
+
+The margin-based exponential loss for classification,
+which penalizes every prediction where `agreement != 1` exponentially.
+It is infinitely many times differentiable, locally Lipschitz continuous
+and strictly convex.
+
+$L(y, ŷ) = exp(-y⋅ŷ)$
+
+"""
+
+immutable ExpLoss <: MarginLoss end
+
+value(loss::ExpLoss, agreement::Number) = exp(-agreement)
+deriv(loss::ExpLoss, agreement::Number) = -exp(-agreement)
+deriv2(loss::ExpLoss, agreement::Number) = exp(-agreement)
+value_deriv(loss::ExpLoss, agreement::Number) = (eᵗ = exp(-agreement); (eᵗ, -eᵗ))
+
+isunivfishercons(::ExpLoss) = true
+isdifferentiable(::ExpLoss) = true
+isdifferentiable(::ExpLoss, at) = true
+istwicedifferentiable(::ExpLoss) = true
+istwicedifferentiable(::ExpLoss, at) = true
+islocallylipschitzcont(::ExpLoss) = true
+islipschitzcont(::ExpLoss) = false
+isconvex(::ExpLoss) = true
+isstrictlyconvex(::ExpLoss) = true
+isstronglyconvex(::ExpLoss) = false
+isclipable(::ExpLoss) = false
+
+
+# ============================================================
+
+doc"""
+    SigmoidLoss <: MarginLoss
+
+Continuous loss which penalizes every prediction in the range (0,2).
+It is infinitely many times differentiable, Lipschitz continuous but nonconvex.
+
+$L(y, ŷ) = 1 - tanh(y⋅ŷ)$
+
+"""
+
+immutable SigmoidLoss <: MarginLoss end
+
+value(loss::SigmoidLoss, agreement::Number) = one(agreement) - tanh(agreement)
+deriv(loss::SigmoidLoss, agreement::Number) = -abs2(sech(agreement))
+deriv2{T<:Number}(loss::SigmoidLoss, agreement::T) = T(2) * tanh(agreement) * abs2(sech(agreement))
+value_deriv(loss::SigmoidLoss, agreement::Number) = (one(agreement) - tanh(agreement), -abs2(sech(agreement)))
+
+isunivfishercons(::SigmoidLoss) = true
+isdifferentiable(::SigmoidLoss) = true
+isdifferentiable(::SigmoidLoss, at) = true
+istwicedifferentiable(::SigmoidLoss) = true
+istwicedifferentiable(::SigmoidLoss, at) = true
+islocallylipschitzcont(::SigmoidLoss) = true
+islipschitzcont(::SigmoidLoss) = true
+isclasscalibrated(::SigmoidLoss) = true
+isconvex(::SigmoidLoss) = false
+isstrictlyconvex(::SigmoidLoss) = false
+isstronglyconvex(::SigmoidLoss) = false
+isclipable(::SigmoidLoss) = false
+
+
+
+# ============================================================
+
+doc"""
+    DWDMarginLoss <: MarginLoss
+
+DWDMarginLoss stands for Distance Weighted Discrimination Margin loss.
+Serves as an alternative to SVM loss function, with similar performance.
+It is Lipschitz continuous and convex, but not strictly convex.
+
+$L(y, ŷ) =  1 - y⋅ŷ                           ... y⋅ŷ ≤ q/(q+1)$
+$L(y, ŷ) = (q^q/(q+1)^(q+1)) / (y⋅ŷ)^q        ... otherwise$
+
+"""
+immutable DWDMarginLoss <: MarginLoss
+    q::Float64
+    function DWDMarginLoss(q::Number)
+        q > 0 || error("q must be strictly positive")
+        new(convert(Float64, q))
+    end
+end
+
+function value{T<:Number}(loss::DWDMarginLoss, agreement::T)
+    q = loss.q
+    agreement <= q/(q+1) ? T(1) - agreement : (q^q/(q+1)^(q+1)) / agreement^q
```

Comment:
good point about the evaluation. Yes, well I wanted to avoid hardcoding Float64 because I am planning on removing the type restriction of the member variable. But I can do that in one go I suppose. I will merge with your proposed solution.